src/p11_child/p11_child_openssl.c

```diff
@@ -0,0 +1,1133 @@
+{
+    bool res = false;
+    int ret;
+    X509_STORE_CTX *ctx = NULL;
+    unsigned long err;
+
+    ctx = X509_STORE_CTX_new();
+    if (ctx == NULL) {
+        err = ERR_get_error();
+        DEBUG(SSSDBG_OP_FAILURE, "X509_STORE_CTX_new failed [%lu][%s].\n",
+                                 err, ERR_error_string(err, NULL));
+        return false;
+    }
+
+    if (!X509_STORE_CTX_init(ctx, p11_ctx->x509_store, cert, NULL)) {
+        err = ERR_get_error();
+        DEBUG(SSSDBG_OP_FAILURE, "X509_STORE_CTX_init failed [%lu][%s].\n",
+                                 err, ERR_error_string(err, NULL));
+        goto done;
+    }
+
+    ret = X509_verify_cert(ctx);
+    if (ret != 1) {
+        DEBUG(SSSDBG_OP_FAILURE, "X509_verify_cert failed [%d].\n", ret);
+        ret = X509_STORE_CTX_get_error(ctx);
+        DEBUG(SSSDBG_OP_FAILURE, "X509_verify_cert failed [%d][%s].\n",
+                                 ret, X509_verify_cert_error_string(ret));
+        ret = EINVAL;
+        goto done;
+    }
+
+    res = true;
+
+done:
+    X509_STORE_CTX_free(ctx);
+
+    return res;
+}
+
+bool do_verification_b64(struct p11_ctx *p11_ctx, const char *cert_b64)
+{
+    int ret;
+    X509 *cert;
+    bool res;
+
+    ret = b64_to_cert(cert_b64, &cert);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to convert certificate.\n");
+        return EINVAL;
+    }
+
+    res = do_verification(p11_ctx, cert);
+    X509_free(cert);
+
+    return res;
+}
+
+
+#if 0
+enum op_mode {
+    OP_NONE,
+    OP_AUTH,
+    OP_PREAUTH,
+    OP_VERIFIY
+};
+
+enum pin_mode {
+    PIN_NONE,
+    PIN_STDIN,
+    PIN_KEYPAD
+};
+#endif
+
+#define ATTR_ID 0
+#define ATTR_LABEL 1
+#define ATTR_CERT 2
+
+struct cert_list {
+    struct cert_list *prev;
+    struct cert_list *next;
+    CK_ATTRIBUTE attributes[3];
+    char *id;
+    char *label;
+    X509 *cert;
+    char *subject_dn;
+    char *cert_b64;
+    CK_KEY_TYPE key_type;
+    CK_OBJECT_HANDLE private_key;
+};
+
+static int free_x509_cert(struct cert_list *item)
+{
+    X509_free(item->cert);
+    return 0;
+}
+
+static int read_certs(TALLOC_CTX *mem_ctx, CK_FUNCTION_LIST *module,
+                      CK_SESSION_HANDLE session, struct p11_ctx *p11_ctx,
+                      struct cert_list **cert_list)
+{
+    int ret;
+    size_t c;
+    CK_RV rv;
+    struct cert_list *list = NULL;
+    struct cert_list *item;
+    X509_NAME *tmp_name;
+    char *tmp_name_str;
+
+    CK_OBJECT_CLASS cert_class = CKO_CERTIFICATE;
+    CK_CERTIFICATE_TYPE cert_type = CKC_X_509;
+    CK_ATTRIBUTE cert_find_template[] = {
+        {CKA_CLASS, &cert_class, sizeof(CK_OBJECT_CLASS)} ,
+        {CKA_CERTIFICATE_TYPE, &cert_type, sizeof(CK_CERTIFICATE_TYPE)}
+    };
+
+    CK_ULONG obj_count;
+    CK_OBJECT_HANDLE obj;
+
+    rv = module->C_FindObjectsInit(session, cert_find_template, 2);
+    if (rv != CKR_OK) {
+        DEBUG(SSSDBG_OP_FAILURE ,"C_FindObjectsInit failed.\n");
```

Comment:
Do you know if these module-> "methods" return useful error codes? If yes, we might want to log them right away or else we risk getting unhelpful log messages from some angry customer a year from now :)

Reply:
Yes, I added the code and the matching error string.